### PR TITLE
raidboss: p8s HC1/HC2 Tower Colors + Merge Targets

### PR DESCRIPTION
**Pull request by Legends0**
Utilizes new MapEffects regex to determine the color of the tower and then existing debuff trigger to determine who to merge with. Will merge Long Debuffs together and Short Debuffs + Stack Players together.

This is a draft for now. Still need to fixup  HC2's non-debuff players and the two color towers.

### CODE REVIEW

**Comment by Legends0** [comment]
The HC2 second towers callout could have a priority for merge, give the player two options, or just state the color of towers.

**Comment by Legends0** [comment]
I realize I also removed the color callout for HC2, but it looks like there is some left over from HC1 that is not properly removed as it would callout two towers for both sets so that too needs to be addressed.

**Comment by Legends0** [comment]
I'm not sure how to keep and/or when to fallback to the hints.

It's possible there are some additional fallback options to add:
- Just call tower colors?
- If MapEffects or return invalid data, possible to add delay to hints triggers and then condition them?

**Comment by quisquous** [comment]
Overall I think this is good.  I think my two biggest issues are:
1) I feel like the big function is hard to follow, with lots of temporary variables and variables being reused (e.g. shortPerfection is both "I am the short a/b/g" and also "I am the perfect a/b/g").  I'm happy to take a crack at rewriting this a little bit more simply if you'd like and send a PR to apply that you can look at and tell me what you think.

2) I disagree with the "who to merge with" text when there are multiple options (e.g. HC1 2nd towers).  It might be worth saying both names?

File: ui/raidboss/data/06-ew/raid/p8s.ts [hunk 1]
```diff
@@ -2064,54 +2086,277 @@ const triggerSet: TriggerSet<Data> = {
...
-          en: 'Purple/Blue Tower',
-          de: 'Lilaner/Blauer Turm',
-          fr: 'Tour Violette/Bleue',
-          ja: '紫・青',
-          ko: '보라/파랑 기둥',
-        },
+      id: 'P8S Arcane Channel Collect',
+      type: 'MapEffect',
+      netRegex: NetRegexes.mapEffect({ flags: arcaneChannelFlags }),
+      condition: (data) => data.seenFirstTankAutos,
```
[the remark this answers is not in the document]
Comment (Legends0) [reply ui/raidboss/data/06-ew/raid/p8s.ts:2125]: I believe the same mapEffect flags exist in phase 1, so I added the condition to prevent further checks on the location value that are in the run section.

File: ui/raidboss/data/06-ew/raid/p8s.ts [hunk 2]
```diff
@@ -2064,54 +2086,277 @@ const triggerSet: TriggerSet<Data> = {
...
+            en: '${color} Tower, Merge with ${player}',
+          },
+          colorTower1MergeLetter: {
+            en: '${color} Tower, Merge with ${letter}',
+          },
+          colorTowerAvoid: {
+            en: 'Avoid ${color} Towers',
+          },
+          colorTowersAvoid: {
+            en: 'Avoid ${color1}/${color2} Towers',
```
[the remark this answers is not in the document]
Comment (Legends0) [reply ui/raidboss/data/06-ew/raid/p8s.ts]: That seems like an easy fix to make. I've implemented this change.

File: ui/raidboss/data/06-ew/raid/p8s.ts [hunk 3]
```diff
@@ -1902,18 +1911,31 @@ const triggerSet: TriggerSet<Data> = {
...
         // 8 and 26s second debuffs.
         const isLong = parseFloat(matches.duration) > 10;
-        if (id === 'D02')
+        if (id === 'D02') {
           data.concept[matches.target] = isLong ? 'longalpha' : 'shortalpha';
-        else if (id === 'D03')
+          if (isLong)
+            data.perfectionLong[matches.target] = 'alpha';
+          else
+            data.perfectionShort[matches.target] = 'alpha';
```
[the remark this answers is not in the document]
Comment (Legends0) [reply ui/raidboss/data/06-ew/raid/p8s.ts:1978]: I think it can be reworked to use just concept, may need to have some extra condition checking for parsing the long/short and handling of splicers.

**Comment by Legends0** [comment]
>     1. I feel like the big function is hard to follow, with lots of temporary variables and variables being reused (e.g. shortPerfection is both "I am the short a/b/g" and also "I am the perfect a/b/g").  I'm happy to take a crack at rewriting this a little bit more simply if you'd like and send a PR to apply that you can look at and tell me what you think.

At one point I had it outputting for every single case by using the same return statement, in effect having three or so returns per tower set. To reduce the duplicate code I used the temporary variables, but this does leave out the option for players to have different text per set.

>     2. I disagree with the "who to merge with" text when there are multiple options (e.g. HC1 2nd towers).  It might be worth saying both names?
I did at one point try out saying both names for High Concept 2, which it could still do but I went with the common PF strategy for both concepts. I imagine there are more people varying High Concept 1 second tower who merges with who than there are High Concept 2 and I do see the benefit of knowing who else you could merge with if the player cactbot calls out went with a different player outside of different priority used.

File: ui/raidboss/data/06-ew/raid/p8s.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Legends0) [reply ui/raidboss/data/06-ew/raid/p8s.ts]: I have removed the data elements of perfectionShort and perfectionLong to just use concept for that, however it still requires some additional rework should a clone not be required to filter for long/short players. This is largely to keep most of the current input/output the same before taking it a step further.

**Comment by Legends0** [comment]
I've added the name of the second player that could be merged with during the second towers of high concept.

As with High Concept 2 it is also possible for merging with two players, should this be updated to reflect that?
Unused `shortalpha` can merge with `longbeta`, and unused `shortbeta` can merge with `longalpha`, leaving the other long player to merge with `longgamma`.

**Comment by Legends0** [comment]
Hmm, seems the npm that is run by the check doesn't support structuredClone. That's odd, this was added in 17, I guess the engine is set to >= 16 prevents this from passing?

**Comment by Legends0** [comment]
@quisquous I am not sure if one of these avoid tower lines never gets called or how they get called

**Comment by Legends0** [comment]
Ok I think I figured out when they get used. I've also changed the `animal` to `primal`.

**Comment by quisquous** [comment]
Looks good to me! If there's other fixups, feel free to send more PRs.  :)

Thank you for this!
